plugins/modules/mso_schema_site_service_graph.py

```diff
@@ -0,0 +1,273 @@
+    template_idx = template_names.index(template)
+
+    # Get site
+    site_id = mso.lookup_site(site)
+
+    # Get site_idx
+    if 'sites' not in schema_obj:
+        mso.fail_json(msg="No site associated with template '{0}'. Associate the site with the template using mso_schema_site.".format(template))
+    sites = [(s.get('siteId'), s.get('templateName')) for s in schema_obj.get('sites')]
+    if (site_id, template) not in sites:
+        mso.fail_json(msg="Provided site-template association '{0}-{1}' does not exist.".format(site, template))
+
+    # Schema-access uses indexes
+    site_idx = sites.index((site_id, template))
+    # Path-based access uses site_id-template
+    site_template = '{0}-{1}'.format(site_id, template)
+
+    mso.existing = {}
+    service_graph_idx = None
+
+    # Get Service Graph
+    service_graph_ref = mso.service_graph_ref(schema_id=schema_id, template=template, service_graph=service_graph)
+    service_graph_refs = [f.get('serviceGraphRef') for f in schema_obj.get('sites')[site_idx]['serviceGraphs']]
+    if service_graph is not None and service_graph_ref in service_graph_refs:
+        service_graph_idx = service_graph_refs.index(service_graph_ref)
+        mso.existing = schema_obj.get('sites')[site_idx]['serviceGraphs'][service_graph_idx]
+
+    if state == 'query':
+        if service_graph is None:
+            mso.existing = schema_obj.get('sites')[site_idx]['serviceGraphs']
+        elif service_graph is not None and service_graph_idx is None:
+            mso.fail_json(msg="Service Graph '{service_graph}' not found".format(service_graph=service_graph))
+        mso.exit_json()
+
+    service_graphs_path = '/sites/{0}/serviceGraphs/-'.format(site_template)
+    service_graph_path = '/sites/{0}/serviceGraphs/{1}'.format(site_template, service_graph)
+    ops = []
+
+    mso.previous = mso.existing
+    if state == 'absent':
+        if mso.existing:
+            mso.sent = mso.existing = {}
+            ops.append(dict(op='remove', path=service_graph_path))
+
+    elif state == 'present':
+        devices_payload = []
+        service_graphs = templates[template_idx]['serviceGraphs']
+        for graph in service_graphs:
+            if graph.get('name') == service_graph:
+                service_node_types_from_template = graph['serviceNodes']
+        user_number_devices = len(devices)
+        number_of_nodes_in_template = len(service_node_types_from_template)
+        if user_number_devices != number_of_nodes_in_template:
+            mso.fail_json(msg="Service Graph '{0}' has '{1}' service node type(s) but '{2}' service node type(s) were given for the device"
+                          .format(service_graph, number_of_nodes_in_template, user_number_devices))
+
+        if devices is not None:
+            for index, device in enumerate(devices):
+                service_node_type_names_from_template = [type.get('name') for type in service_node_types_from_template]
```

Comment:
Not here. The line between if and for loop. You only need this executed once.

Reply:
yeah